### PR TITLE
SDFD-32: feat ledger module

### PR DESCRIPTION
**Pull request by lccoronel**
##Ref for Operations Capacity: [Ledger module · Issue #275 · stellar/dashboard](https://github.com/stellar/dashboard/issues/275#issuecomment-1147795658) 

### Considerations:

* Trendline is not necessary to implement;

* The number of red line It's 518,400,000.

* The "Ledger" container component will need to be created

* Use [Stellar Design System](https://design-system.stellar.org/) (SDS) components as much as possible.

* Icons should also be used from the SDS ( component). It's OK if they differ slightly.

* Retrieve data for calculations from SDK

* On Mainnet (and not TestNet)

Screen Shot 2022-07-25 at 09 27 40

### CODE REVIEW

File: frontend/src/ducks/ledgers.ts [hunk 1]
```diff
@@ -141,6 +143,32 @@ export const fetchLedgersTransactionsHistoryAction = createAsyncThunk<
   },
 );
 
+export const fetchLedgersModule = createAsyncThunk<
```
Comment (acharb) [inline frontend/src/ducks/ledgers.ts:146]: lets make the name more general, eg. `fetchLedgerOperations`

File: frontend/src/components/LedgerModule/index.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import "./styles.scss";
+
+import { fetchLedgersModule } from "ducks/ledgers";
+import { SectionCard } from "components/SectionCard";
+import { VerticalBarChart } from "components/charts/VerticalBarChart";
+import { useRedux } from "hooks/useRedux";
+import { ActionStatus } from "types";
+
+const limitLine = 518400000;
```
Comment (acharb) [inline frontend/src/components/LedgerModule/index.tsx:12]: let's add a comment here for where this number is coming from
eg

```
// max operations per month = 12*1000*60*24*30 = 518400000
```

File: frontend/package.json [hunk 3]
```diff
@@ -100,6 +100,7 @@
     "html-webpack-plugin": "^5.5.0",
     "mini-css-extract-plugin": "^2.6.0",
     "prettier": "^2.5.1",
+    "sass": "^1.53.0",
```
Comment (acharb) [inline frontend/package.json:103]: just curious what's this needed for?

File: frontend/src/components/LedgerModule/index.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  useEffect(() => {
+    dispatch(fetchLedgersModule());
+  }, [dispatch]);
+
+  return (
+    <SectionCard
+      title="Ledger Module"
+      titleLinkLabel="API"
+      titleLink={`/api/ledgers`}
```
Comment (acharb) [inline frontend/src/components/LedgerModule/index.tsx:26]: this titleLink is incorrect, should be `/api/ledgers/hour/public` I think

File: frontend/src/ducks/ledgers.ts [hunk 5]
```diff
@@ -141,6 +143,32 @@ export const fetchLedgersTransactionsHistoryAction = createAsyncThunk<
...
+    const response = await fetch(
+      `/api/ledgers${historyFilter.endpointPrefix}${
+        networkConfig[Network.MAINNET].ledgerTransactionsHistorySuffix
+      }`,
+    );
+
+    const operations = await response.json();
+
+    const result = operations.data.map((operation: LedgerModuleItem) => ({
+      date: new Date(operation.end),
```
Comment (acharb) [inline frontend/src/ducks/ledgers.ts:162]: we should use a serializable data for this. If you look at the console logs this throws an error:
```
react_devtools_backend.js:4026 A non-serializable value was detected in the state, in the path: `ledgers.ledgerModule.0.date`. Value: Thu Jun 23 2022 09:38:06 GMT-0700 (Pacific Daylight Time) 
```

File: frontend/package.json (hunk 3 above)
Comment (lccoronel) [reply frontend/package.json:103]: because `node-sass` lib was depreciated and we can’t use on machines with M1 chip, so when use `sass` everything works 😊

File: frontend/src/ducks/ledgers.ts [hunk 6]
```diff
@@ -141,6 +143,35 @@ export const fetchLedgersTransactionsHistoryAction = createAsyncThunk<
...
+    const response = await fetch(
+      `/api/ledgers${historyFilter.endpointPrefix}${
+        networkConfig[Network.MAINNET].ledgerTransactionsHistorySuffix
+      }`,
+    );
+
+    const operations = await response.json();
+
+    const result = operations.data.map((operation: LedgerModuleItem) => ({
+      date: parseDateFromFormat(
```
Comment (acharb) [inline frontend/src/ducks/ledgers.ts:162]: this `fetchLedgerOperations` call is failing with error `Invalid time value`, probably the parsing is incorrect somehow

Comment (lccoronel) [reply frontend/src/ducks/ledgers.ts]: 😬 sorry I didn't see that

File: frontend/src/components/LedgerOperations/index.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        networkConfig[Network.MAINNET].ledgerTransactionsHistorySuffix
+      }`}
+      isLoading={ledgers.status === ActionStatus.PENDING}
+      noData={!ledgers.ledgerOperations.length}
+    >
+      {ledgers.ledgerOperations.length > 0 && (
+        <div className="LedgerOperations__mainChart">
+          <div className="LedgerOperations__mainChart__container">
+            <VerticalBarChart
+              data={data}
```
Comment (acharb) [inline frontend/src/components/LedgerOperations/index.tsx:50]: data dates look backwards, I think we should sort before passing here

File: frontend/src/components/LedgerOperations/index.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import { fetchLedgerOperations } from "ducks/ledgers";
+import { SectionCard } from "components/SectionCard";
+import { VerticalBarChart } from "components/charts/VerticalBarChart";
+import {
+  ledgerTransactionHistoryConfig,
+  networkConfig,
+} from "constants/settings";
+
+// max operations per month = 12*1000*60*24*30 = 518400000
+const limitLine = 518400000;
```
Comment (acharb) [inline frontend/src/components/LedgerOperations/index.tsx:17]: actually we should lower this number to the max per day, not per month, since that's what this chart is showing
(@jcx120 correct me if I'm wrong)

so instead it would be
```
// max operations per day = 12*60*24*1000 = 17,280,000
const limitLine = 17280000;
```

Comment (jcx120) [reply frontend/src/components/LedgerOperations/index.tsx:17]: The Max ops line is per month. The X axis is currently shown in the staging instance as daily but should be monthly per the original example in the [doc](https://docs.google.com/document/d/1J2HruGsFGluNSY6IXJ9ooSnIYW5gv_LM-udYnIkkoaw/edit).

**Comment by acharb** [comment]
hey @lccoronel we actually need the chart to be per month. See Jesse's comment here: https://github.com/stellar/dashboard/pull/295#discussion_r922365375

So the red line should go back to `518400000` (that was changed in the last [commit](https://github.com/stellar/dashboard/pull/295/commits/d493df7738598900585d43d0fc951ba29add8fa4))

And should look similar to (chart in the [docs](https://docs.google.com/document/d/1J2HruGsFGluNSY6IXJ9ooSnIYW5gv_LM-udYnIkkoaw/edit)):

Screen Shot 2022-07-18 at 12 08 20 PM


I think this will probably involve adding a new backend endpoint, as from I know we don't have this data available. So we'll have to grab it from Big Query cc @Selhar

**Comment by Selhar** [comment]
What data do we need? Currently ledgers endpoint returns on a monthly, daily and hourly interval. @acharb

**Comment by Selhar** [comment]
Hey @acharb, just to confirm, this cache will take potentially more than a day to generate, given the amount of data we're bringing in, is this ok? We'll only have to generate it once in production though, but it will take a really long time.

**Comment by acharb** [comment]
@Selhar ya  I think that should be fine 🤔 we could probably run as infrequent as once a month.

What's the query you're running?
(or I can take a look whenever the PR comes out)

**Comment by Selhar** [comment]
Oh it's the same query we already use for current ledgers, the only change is the filter goes to january 1 2018.

**Comment by acharb** [comment]
ah, I think we should just grab the operations/month from big query, instead of "catching up" like we do with the current backend ledger endpoints

File: frontend/src/components/RecentOperations/index.tsx [hunk 9]
```diff
@@ -12,7 +12,11 @@ import { OperationTypeColumn } from "./OperationTypeColumn";
...
 
-export const RecentOperations = () => {
+export const RecentOperations = ({
+  network = Network.MAINNET,
+}: {
+  network?: Network;
+}) => {
   const { operations } = useRedux("operations");
   const dispatch = useDispatch();
   const horizonURL = "https://horizon.stellar.org";
```
Comment (quietbits) [inline frontend/src/components/RecentOperations/index.tsx:22]: Not part of this PR, why do we hard code Horizon URL here instead of getting it from `networkConfig` in `constants/settings.ts`?

File: frontend/src/ducks/ledgers.ts [hunk 10]
```diff
@@ -141,6 +143,32 @@ export const fetchLedgersTransactionsHistoryAction = createAsyncThunk<
   },
 );
 
+export const fetchLedgerOperations = createAsyncThunk<
+  FetchLedgerOperationsResponse[],
+  void,
```
Comment (quietbits) [inline frontend/src/ducks/ledgers.ts:148]: Should we use `undefined` instead of `void` here for the accepted parameter (or not param in this case)? `void` feels more like a type for a function that doesn't return anything. I'm not very opinionated about it, just curious what you think. :slightly_smiling_face:

**Review by quietbits — COMMENTED** [review]
The website breaks when switching to the Testnet tab. We might want to add `id` check in the `AccountBadge` component to make sure it won't crash the website. It might be good to double-check the data as well, just to make sure it's OK that it doesn't return the `id` value we're expecting.

image

**Comment by lccoronel** [comment]
@acharb @quietbits  @jcx120  

After @Selhar merged the [ledgers operations PR](https://github.com/stellar/dashboard/pull/297), we start to change the ledgers operation chart and it was like this:

Screen Shot 2022-07-22 at 08 57 29

except for the chart bars, the rest of the chart data is not yet integrated with the API, but you can already have a preview, is it as you imagined?

**Comment by Selhar** [comment]
i think this data is in the opposite direction, isn't it? As in, 2022 is to the left, and 2015 to the right.

File: frontend/src/components/LedgerOperations/index.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    <SectionCard
+      title="Ledger Operations"
+      titleLinkLabel="API"
+      titleLink={`/api/ledgers${historyFilter.endpointPrefix}${
+        networkConfig[Network.MAINNET].ledgerTransactionsHistorySuffix
+      }`}
+      isLoading={ledgers.status === ActionStatus.PENDING}
+      noData={!ledgers.ledgerOperations.length}
+    >
+      {ledgers.ledgerOperations.length > 0 && (
```
[the remark this answers is not in the document]
Comment (lccoronel) [reply frontend/src/components/LedgerOperations/index.tsx:48]: Actually I think that we don’t need this validation more 😁

File: frontend/src/components/RecentOperations/index.tsx (hunk 9 above)
Comment (lccoronel) [reply frontend/src/components/RecentOperations/index.tsx:22]: sorry, I took as base the dashboards v1, I will fix it

**Comment by lccoronel** [comment]
> i think this data is in the opposite direction, isn't it? As in, 2022 is to the left, and 2015 to the right.

Yes, I just put the data on the chart to show how it would look with a lot of data, I haven't organized it yet

File: frontend/src/ducks/ledgers.ts (hunk 10 above)
Comment (lccoronel) [reply frontend/src/ducks/ledgers.ts:148]: I prefer use `undefined`, but I left `void` following the design pattern, in some other functions they were already like this.

Comment (quietbits) [reply frontend/src/ducks/ledgers.ts:148]: Let's use `undefined`, we prefer it as well. We must have missed `void` :sweat_smile: . Thank you!

**Comment by acharb** [comment]
besides the backwards data the charts looking good to me

we need to change the "max" line back to `518,400,000` though, since now we're doing per month and not per day

**Review by quietbits — APPROVED** [review]
